Add Makefile, use golangci-lint, move to GitHub actions

guggero opened the pull request:
As a new contributor to the Neutrino repository it was a bit of a challenge to figure out how to use the clunky `gotest.sh` file. That's why I think it would be nice to switch over to a `Makefile`.

At the same time I felt like we should fix a number of smaller issues that an updated linter should find. That's why I opted to add the `golangci-lint` linter as well.
And I think I might even have discovered a possible explanation for some of the flakes we see in the `lnd` itests with Neutrino (see 0a824919, could also be mistaken about the significance of this).

wpaulino commented on blockntfns/manager.go:
```diff
@@ -139,6 +139,7 @@ func (m *SubscriptionManager) Stop() {
 	var wg sync.WaitGroup
 	wg.Add(len(m.subscribers))
 	for _, subscriber := range m.subscribers {
+		subscriber := subscriber
```
Nit: could also provide it as an argument to the goroutine func.

wpaulino commented on batch_spend_reporter.go:
```diff
@@ -61,14 +61,16 @@ func (b *batchSpendReporter) NotifyUnspentAndUnfound() {
 		}
 
-		b.notifyRequests(&outpoint, requests, tx, nil)
+		b.notifyRequests(&op, requests, tx, nil)
```
:+1:

wpaulino commented on blockmanager.go:
```diff
@@ -803,7 +803,7 @@ func (b *blockManager) getUncheckpointedCFHeaders(
 	pristineHeaders, ok := headers[key]
 	if !ok {
-		return fmt.Errorf("All peers served bogus headers! Retrying " +
+		return fmt.Errorf("all peers served bogus headers! Retrying " +
```
Nit: replace `!` with a comma?

wpaulino commented on neutrino.go:
```diff
@@ -1489,9 +1489,13 @@ func (s *ChainService) Start() error {
 	s.blockManager.Start()
 	s.blockSubscriptionMgr.Start()
-	s.workManager.Start()
+	if err := s.workManager.Start(); err != nil {
```
:+1:

wpaulino commented on blockmanager_test.go:
```diff
@@ -440,7 +440,7 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 			startHeight = wire.CFCheckptInterval / 3
 		}
-		go func() {
+		go func() { // nolint:staticcheck
```
Can we remove the lint exception for all the `t.Fatal` in goroutines and follow https://golang.org/doc/go1.16#vet?

guggero replied:
Ah, I didn't know it was that easy to fix (expected a bigger refactor). Thanks for the hint, done!

wpaulino commented on headerfs/store.go:
```diff
@@ -85,14 +85,14 @@ var headerBufPool = sync.Pool{
 // TODO(roasbeef): quickcheck coverage
 type headerStore struct {
-	mtx sync.RWMutex
+	mtx sync.RWMutex // nolint:structcheck, false positive
```
Interesting. I removed the lint exception to see what the warning was and got `mtx is unused`. It seems to not like the fact that `headerStore` is embedded in other structs.

guggero replied:
Yeah, this is a bit annoying. I tried the newest version of `golanglint-ci` but that also still flags it. Added a comment to make clear why it's a false positive.

wpaulino commented on neutrino.go:
```diff
@@ -158,7 +158,7 @@ type ServerPeer struct {
 	persistent     bool
 	knownAddresses map[string]struct{}
-	banScore       connmgr.DynamicBanScore
+	banScore       connmgr.DynamicBanScore // nolint:unused, false positive
```
The ban scoring logic was removed in a previous commit, this should probably be removed there as well.

guggero replied:
Right, removed.

wpaulino commented on pushtx/broadcaster.go:
```diff
@@ -81,19 +81,19 @@ func NewBroadcaster(cfg *Config) *Broadcaster {
 // carrying out its duties.
 func (b *Broadcaster) Start() error {
-	var err error
+	var returnErr error
```
:+1:

wpaulino commented on query/workmanager_test.go:
```diff
@@ -202,13 +202,13 @@ func TestWorkManagerWorkDispatcherFailures(t *testing.T) {
 		wk := workers[i]
 		go func() {
-			for {
+			for { // nolint:gosimple
```
Looks like we can remove the `select` since we're only waiting on one channel?

wpaulino reviewed (commented):
Neutrino was clearly in need of a better linter! `btcwallet` would be another good candidate.

It looks like the GitHub actions aren't running yet, is there anything else that needs to be done?

guggero commented:
> It looks like the GitHub actions aren't running yet, is there anything else that needs to be done?

Changes in GH actions in a PR aren't applied until the PR is merged. I think it's a security measure to make sure repository secrets cannot be extracted easily just by opening a PR.

guggero commented:
@cfromknecht PTAL

halseth commented on .golangci.yml:
```diff
@@ -0,0 +1,51 @@
+    - lll
+
+    # We don't care (enough) about misaligned structs to lint that.
+    - maligned
```
maybe some of these could be enabled?

guggero replied:
Yeah, why not. I enabled `malign` globally but only fixed one struct that is used often. For the others the readability/order of the struct would suffer too much for very little benefit.

Also skipping the `prealloc` check only for test code.

halseth commented on batch_spend_reporter.go:
```diff
@@ -61,14 +61,16 @@ func (b *batchSpendReporter) NotifyUnspentAndUnfound() {
 	log.Debugf("Finished batch, %d unspent outpoints", len(b.requests))
 
 	for outpoint, requests := range b.requests {
+		op := outpoint
```
yeah, good to catch these! 😀

halseth reviewed (approved):
Excellent! Awesome to finally bring neutrino into the new decade 🚀

halseth commented:
@guggero Feel free to merge when you've checked out whether the comments are actionable

guggero commented:
@halseth I rebased and added two more commits. I don't have merge permissions in this repo so feel free to push the button if you think this is ready.

halseth commented on .golangci.yml:
```diff
@@ -48,3 +44,7 @@ issues:
+
+        # Instances of table driven tests that don't pre-allocate shouldn't
+        # trigger the linter.
+        - prealloc
```
👍
